# Домашнее задание к уроку 1: Основы PyTorch

**Цель задания**

**Задание 1: Создание и манипуляции с тензорами (25 баллов)**


**1.1 Создание тензоров (7 баллов)**


In [2]:
import torch
tensor1 = torch.rand(3, 4)
tensor2 = torch.zeros(2, 3, 4)
tensor3 = torch.ones(5, 5)
tensor4 = torch.arange(16).reshape(4, 4)
print(tensor1)
print(tensor2)
print(tensor3)
print(tensor4)

tensor([[0.3875, 0.5233, 0.9148, 0.6759],
        [0.8672, 0.9119, 0.8236, 0.6304],
        [0.3133, 0.0824, 0.1708, 0.4995]])
tensor([[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]])
tensor([[1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.]])
tensor([[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11],
        [12, 13, 14, 15]])


**1.2 Операции с тензорами (6 баллов)**

In [6]:
A = torch.rand(3, 4)
B = torch.rand(4, 3)
T_A = A.T #Транспонируем
dot = A @ B #Матричное умножение
mul = A * B.T #Поэлементное умножение
sm = A.sum() #Сумма всех элементов
print(A)
print(B)
print(T_A)
print(dot)
print(mul)
print(sm)

tensor([[0.8978, 0.1595, 0.3566, 0.3068],
        [0.7522, 0.1738, 0.1157, 0.3482],
        [0.9595, 0.5632, 0.6350, 0.5358]])
tensor([[0.8844, 0.7155, 0.7917],
        [0.3481, 0.7835, 0.8502],
        [0.9246, 0.6858, 0.8403],
        [0.7907, 0.8700, 0.1376]])
tensor([[0.8978, 0.7522, 0.9595],
        [0.1595, 0.1738, 0.5632],
        [0.3566, 0.1157, 0.6350],
        [0.3068, 0.3482, 0.5358]])
tensor([[1.4218, 1.2788, 1.1883],
        [1.1080, 1.0566, 0.8884],
        [2.0554, 2.0294, 1.8458]])
tensor([[0.7940, 0.0555, 0.3297, 0.2426],
        [0.5382, 0.1362, 0.0793, 0.3029],
        [0.7597, 0.4788, 0.5336, 0.0737]])
tensor(5.8041)


**1.3 Индексация и срезы (6 баллов)**

In [11]:
tensor = torch.arange(125).reshape(5, 5, 5) #матрица 5x5x5 от 0 до 124
first_row = tensor[0] #берем первую строку
last_col = tensor[:, :, -1] #берем последний столбец
cntr = tensor[2, 1:3, 1:3] #берем Подматрицу размером 2x2 из центра тензора
even = tensor[::2, ::2, ::2] #берем Все элементы с четными индексами
print(first_row)
print(last_col)
print(cntr)
print(even)

tensor([[ 0,  1,  2,  3,  4],
        [ 5,  6,  7,  8,  9],
        [10, 11, 12, 13, 14],
        [15, 16, 17, 18, 19],
        [20, 21, 22, 23, 24]])
tensor([[  4,   9,  14,  19,  24],
        [ 29,  34,  39,  44,  49],
        [ 54,  59,  64,  69,  74],
        [ 79,  84,  89,  94,  99],
        [104, 109, 114, 119, 124]])
tensor([[56, 57],
        [61, 62]])
tensor([[[  0,   2,   4],
         [ 10,  12,  14],
         [ 20,  22,  24]],

        [[ 50,  52,  54],
         [ 60,  62,  64],
         [ 70,  72,  74]],

        [[100, 102, 104],
         [110, 112, 114],
         [120, 122, 124]]])


**1.4 Работа с формами (6 баллов)**

In [10]:
ten = torch.randn(24)
nten1= ten.reshape(2, 12) #изменяем размер матрицы на 2x12
nten2 = ten.reshape(3, 8)
nten3  = ten.reshape(4, 6)
nten4 = ten.reshape(2, 3, 4)
nten5 = ten.reshape(2, 2, 2, 3)
print(nten1)
print(nten2)
print(nten3)
print(nten4)
print(nten5)

tensor([[-0.4246,  0.3795,  0.3515, -0.3642,  0.7468, -0.3539,  1.4591,  0.2743,
          0.1225,  2.4178,  0.8515,  1.4474],
        [ 1.4324, -0.5123, -1.7494,  0.0270, -0.0557,  0.5359, -0.6298,  1.1441,
         -0.1826,  0.6526, -1.7132, -1.2998]])
tensor([[-0.4246,  0.3795,  0.3515, -0.3642,  0.7468, -0.3539,  1.4591,  0.2743],
        [ 0.1225,  2.4178,  0.8515,  1.4474,  1.4324, -0.5123, -1.7494,  0.0270],
        [-0.0557,  0.5359, -0.6298,  1.1441, -0.1826,  0.6526, -1.7132, -1.2998]])
tensor([[-0.4246,  0.3795,  0.3515, -0.3642,  0.7468, -0.3539],
        [ 1.4591,  0.2743,  0.1225,  2.4178,  0.8515,  1.4474],
        [ 1.4324, -0.5123, -1.7494,  0.0270, -0.0557,  0.5359],
        [-0.6298,  1.1441, -0.1826,  0.6526, -1.7132, -1.2998]])
tensor([[[-0.4246,  0.3795,  0.3515, -0.3642],
         [ 0.7468, -0.3539,  1.4591,  0.2743],
         [ 0.1225,  2.4178,  0.8515,  1.4474]],

        [[ 1.4324, -0.5123, -1.7494,  0.0270],
         [-0.0557,  0.5359, -0.6298,  1.1441],
    

**Задание 2: Автоматическое дифференцирование (25 баллов)** \
**2.1 Простые вычисления с градиентами (8 баллов)**

In [5]:
x = torch.rand(2, requires_grad=True)
y = torch.rand(2, requires_grad=True)
z = torch.rand(2, requires_grad=True)
f = (x**2 + y**2 + z**2 + 2*x*y*z).sum()
f.backward()
print(f'x {x}')
print(f'y {y}')
print(f'z {z}')
print(x.grad)
print(y.grad)
print(z.grad)

x tensor([0.9407, 0.4009], requires_grad=True)
y tensor([0.4875, 0.9949], requires_grad=True)
z tensor([0.1093, 0.0363], requires_grad=True)
tensor([1.9879, 0.8740])
tensor([1.1805, 2.0190])
tensor([1.1357, 0.8704])


**Проверка результата** \
f(x,y,z) = x^2 + y^2 + z^2 + 2*x*y*z \
df/dx = 2x + 0 + 0 + 2yz = 2x + 2yz \
df/dy = 2y + 2xz \
df/dz = 2z + 2xy \
Подставим: \
**df/dx** = 2 * [0.9407, 0.4009] + 2 * [0.4875, 0.9949] * [0.1093, 0.0363] = [1.8814, 0.8018] + 2 * [0.0533, 0.0361] = \
 = [1.8814, 0.8018] + [0.1066, 0.0722] = **[1.9880, 0.8740]** \
**df/dy** = 2 * [0.4875, 0.9949] + 2 * [0.9407, 0.4009] * [0.1093, 0.0363] = [0.9750, 1.9898] + 2 * [0.1029, 0.0146] = \
 = [0.9750, 1.9898] + [0.2058, 0.0292] = **[1.1808, 2.0190]** \
**df/dz** = 2 * [0.1093, 0.0363] + 2 * [1.9879, 0.8740] * [1.1805, 2.0190] = [0.2186, 0.0726] + 2 * [0.4585, 0.3989] = \
 = [0.2186, 0.0726] + [0.9170, 0.7978] = **[1.1356, 0.8704]** \
Результаты одинаковы.

**2.2 Градиент функции потерь (9 баллов)**

In [12]:
x = torch.randint(1, 10, (2,), dtype=torch.float32, requires_grad=True) #тензор x из случайных целых чисел от 1 до 9, размерностью (2,)
y_true = torch.rand(2, requires_grad=True) #тензор y_true со случайными значениями от 0 до 1, размерностью (2,)
w = torch.tensor(1.0, requires_grad=True)
b = torch.tensor(2.0, requires_grad=True)
y_pred = w * x + b
# MSE = (1/n) * Σ(y_pred - y_true)^2
mse = ((y_pred - y_true) ** 2).sum()
mse.backward()
print(w.grad)
print(b.grad)

tensor(112.4291)
tensor(22.9556)


**2.3 Цепное правило (8 баллов)**

In [15]:
x = torch.tensor(1.0, requires_grad=True) #создаем тензор
f = torch.sin(x**2 + 1) #высчитываем функцию
f.backward()
print(x.grad)
x.grad.zero_() #очищаем тензор
f = torch.sin(x**2 + 1) #заново высчитываем функцию
grad = torch.autograd.grad(f, x)
print(grad)

tensor(-0.8323)
(tensor(-0.8323),)


**Ручная проверка** \
**df/dx** = 2x * cos(x^2+1) = 1*cos(2) = 2 * (-0.4161) = -0.8322

**Задание 3: Сравнение производительности CPU vs CUDA (20 баллов)** \
**3.1 Подготовка данных (5 баллов)**

In [16]:
tensor1 = torch.rand(64, 1024, 1024)# Матрица размером 64 × 1024 × 1024, заполненная случайными числами от 0 до 1
tensor2 = torch.rand(128, 512, 512)
tensor3 = torch.rand(256, 256, 256)

**3.2 Функция измерения времени (5 баллов)**

In [19]:
import time

def time_cpu(func, *args):
    """
    Измеряет время выполнения операции на CPU
    """
    start = time.time() #начала выполнения функции
    func(*args)  #вызываем функцию
    end = time.time() #время окончания
    return end - start
def time_gpu(func, *args):
    """
    Измеряет время выполнения операции на GPU 
    """
    start = torch.cuda.Event(enable_timing=True)
    end = torch.cuda.Event(enable_timing=True)
    
    start.record() #начала выполнения функции
    func(*args) #вызываем функцию
    end.record() #фиксируем момент сразу ПОСЛЕ вызова операции
    
    torch.cuda.synchronize()  # ждём завершения всех операций на GPU
    return start.elapsed_time(end) / 1000  # возвращаем время в секундах

**3.3 Сравнение операций (10 баллов)**

In [26]:
tensor1 = torch.rand(64, 1024, 1024)#поменял размеры тензоров, потому что невозможно умножение двух матриц, где столбец первой не соответствует строке второй и тд
tensor2 = torch.rand(64, 1024, 1024)    
tensor3 = torch.rand(64, 1024, 1024)    
def matmul_op(a, b):
    """
    Матричное умножение
    """
    return torch.matmul(a,b)

def add_op(a, b):
    """
    Поэлементное сложение
    """
    return a + b

def mul_op(a, b):
    """
    Поэлементное умножение
    """
    return a * b

def transpose_op(a):
    """
    Транспонирование
    """
    return a.T

def sum_op(a):
    """
    Вычисление суммы всех элементов
    """
    return a.sum()

device = 'CPU'

results_CPU = {
    "Мат. умножение": time_cpu(matmul_op, tensor1, tensor2),
    "Сложение": time_cpu(add_op, tensor2, tensor3),
    "Поэлементное умножение": time_cpu(mul_op, tensor2, tensor3),
    "Транспонирование": time_cpu(transpose_op, tensor1),
    "Сумма элементов": time_cpu(sum_op, tensor3),
}

device = 'cuda'
t1 = tensor1.to(device)
t2 = tensor2.to(device)
t3 = tensor3.to(device)

results_GPU = {
    "Мат. умножение": time_gpu(matmul_op, t1, t2),
    "Сложение": time_gpu(add_op, t2, t3),
    "Поэлементное умножение": time_gpu(mul_op, t2, t3),
    "Транспонирование": time_gpu(transpose_op, t1),
    "Сумма элементов": time_gpu(sum_op, t3),
}
print(f"{'Операция':<25} | {'CPU (мс)':>10} | {'GPU (мс)':>10} | {'Ускорение':>10}")
print("-" * 60)
for key in results_CPU:
    cpu_time_ms = results_CPU[key] * 1000
    gpu_time_ms = results_GPU[key] * 1000
    speedup = cpu_time_ms / gpu_time_ms
    print(f"{key:<25} | {cpu_time_ms:10.2f} | {gpu_time_ms:10.2f} | {speedup:10.2f}x")



Операция                  |   CPU (мс) |   GPU (мс) |  Ускорение
------------------------------------------------------------
Мат. умножение            |     698.88 |      92.24 |       7.58x
Сложение                  |      58.08 |       5.92 |       9.81x
Поэлементное умножение    |      40.95 |       5.89 |       6.95x
Транспонирование          |       0.04 |       0.00 |      19.21x
Сумма элементов           |       9.89 |       1.54 |       6.42x
